File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/server/RequestHandler.java
package lunalauf.rms.utilities.network.server;

import com.google.gson.JsonParseException;
import javafx.application.Platform;
import javafx.concurrent.Task;
import lunalauf.rms.utilities.network.communication.ErrorType;
import lunalauf.rms.utilities.network.communication.MessageProcessor;
import lunalauf.rms.utilities.network.communication.message.Message;
import lunalauf.rms.utilities.network.communication.message.request.Request;
import lunalauf.rms.utilities.network.communication.message.response.Response;
import lunalauf.rms.utilities.network.communication.message.response.ResponseFactory;

import java.util.concurrent.FutureTask;

public class RequestHandler extends Task<Integer> {

    private final Client client;
    private final LunaLaufAPI api;
    private final ResponseFactory responseFactory;

    public RequestHandler(Client client, LunaLaufAPI api) {
        this.client = client;
        this.api = api;
        this.responseFactory = new ResponseFactory();
    }

    @Override
    protected Integer call() throws Exception {
        updateValue(0);

        try {
            executeProtocol();
        } catch (Exception ignored) {
        }

        updateValue(-1);
        return -1;
    }

    private void executeProtocol() throws Exception {
        updateValue(1);

        String messageString;
        while ((messageString = client.receive()) != null) {
            Response response = null;

            Message message = null;
            try {
                message = MessageProcessor.fromJsonString(messageString);
            } catch (JsonParseException e) {
                response = responseFactory.createErrorResponse(-1, ErrorType.CORRUPTED_CLIENT_MESSAGE);
            }

            if (message != null) {
                if (message instanceof Request request) {
                    FutureTask<Response> task = new FutureTask<>(new ModelUpdater(api, request, responseFactory));
                    Platform.runLater(task);
                    try {
                        response = task.get();
                    } catch (Exception e) {
                        response = responseFactory.createErrorResponse(message.messageId, ErrorType.BAD_SERVER_STATE);
                    }
                } else {
                    response = responseFactory.createErrorResponse(message.messageId, ErrorType.UNEXPECTED_CLIENT_MESSAGE);
                }
            }

            client.send(MessageProcessor.toJsonString(response));

            if (isCancelled())
                break;
        }
    }

}

File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/server/RequestHandler.kt
package lunalauf.rms.utilities.network.server

import com.google.gson.JsonParseException
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import lunalauf.rms.modelapi.ModelState
import lunalauf.rms.utilities.network.communication.ErrorType
import lunalauf.rms.utilities.network.communication.MessageProcessor
import lunalauf.rms.utilities.network.communication.message.request.Request
import lunalauf.rms.utilities.network.communication.message.response.ResponseFactory
import org.slf4j.LoggerFactory

class RequestHandler(
    private val client: Client,
    private val modelState: ModelState,
    private val status: MutableStateFlow<Int>,
    scope: CoroutineScope
) {
    private val logger = LoggerFactory.getLogger(this::class.java)

    private val responseFactory = ResponseFactory()
    private val job: Job

    init {
        status.value = 0
        job = scope.launch {
            try {
                executeProtocol()
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                logger.error("Exception occurred while running the protocol", e)
            } finally {
                status.value = -1
            }
        }
    }

    private suspend fun CoroutineScope.executeProtocol() {
        status.value = 1
        while (true) {
            if (!this.isActive)
                return

            val messageString = client.receive()
                ?: return

            val response = try {
                val message = MessageProcessor.fromJsonString(messageString)

                if (message is Request) {
                    if (modelState is ModelState.Loaded) {
                        val modelUpdater = ModelUpdater(modelState, message, responseFactory)
                        try {
                            modelUpdater.run()
                        } catch (e: Exception) {
                            logger.error("Exception occurred while updating the model", e)
                            responseFactory.createErrorResponse(message.messageId, ErrorType.BAD_SERVER_STATE)
                        }
                    } else {
                        logger.warn("Request handler needs available model!")
                        responseFactory.createErrorResponse(message.messageId, ErrorType.NO_MODEL_STATE)
                    }
                } else {
                    logger.warn("Received unexpected client message: {} from: {}", message, client.remoteAddress)
                    responseFactory.createErrorResponse(message.messageId, ErrorType.UNEXPECTED_CLIENT_MESSAGE)
                }
            } catch (_: JsonParseException) {
                logger.warn("Received corrupted client message from: {}", client.remoteAddress)
                responseFactory.createErrorResponse(-1, ErrorType.CORRUPTED_CLIENT_MESSAGE)
            }

            client.send(MessageProcessor.toJsonString(response))
            logger.info("Response sent: {} to: {}", response, client.remoteAddress)
        }
    }

    fun cancel() {
        job.cancel()
    }
}
